File: Construct Binary Tree from Inorder and Postorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder.length==0) return null;
        int indexpost=postorder.length-1;
        int indexin=inorder.length-1;
        Stack<TreeNode> parent=new Stack<TreeNode>();
        TreeNode root=new TreeNode(postorder[indexpost]);
        indexpost--;
        TreeNode cur=root;
        while(indexpost>=0){
            while(cur.val!=inorder[indexin]){
                parent.push(cur);
                cur.right=new TreeNode(postorder[indexpost]);
                indexpost--;
                cur=cur.right;
            }
            if(indexpost<0) break;
            indexin--;
            while(parent.size()>0&&parent.peek().val==inorder[indexin]){
                indexin--;
                cur=parent.pop();
            }
            cur.left=new TreeNode(postorder[indexpost]);
            indexpost--;
            cur=cur.left;
        }
        return root;
    }
}

File: Construct Binary Tree from Inorder and Postorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param inorder, a list of integers
    # @param postorder, a list of integers
    # @return a tree node
    def buildTree(self, inorder, postorder):
        if len(inorder) == 0:
            return None

        postorder_idx = len(postorder)-1
        inorder_idx = len(inorder)-1

    # not strictly a parent stack
    # more of a stack of nodes whose values haven't appeared in
    # inorder yet
        parent = []

    # set initial node
        root = TreeNode(postorder[postorder_idx])
        postorder_idx -= 1

        cur = root
        while postorder_idx >= 0:
        # if we haven't seen the current value in the inorder array
        # then we know that the next node is going to be the right child
        # of the current node
            while cur.val != inorder[inorder_idx]:
                parent.append(cur)
                cur.right = TreeNode(postorder[postorder_idx])
                postorder_idx -= 1
                cur = cur.right

            if postorder_idx < 0:
                break

            inorder_idx -= 1

        # the next root belongs to a left subtree
        # find the correct subtree it belongs to
            while len(parent) > 0 and parent[-1].val == inorder[inorder_idx]:
                inorder_idx -= 1
                cur = parent.pop()

        # finally set the left subtree and continue the reconstruction
        # from there
            cur.left = TreeNode(postorder[postorder_idx])
            postorder_idx -= 1
            cur = cur.left
        return root
